// telegramsjs/dist/src/structures/misc/SharedUser.js
"use strict";
const { Base } = require("../Base");
const { Photo } = require("../media/Photo");
const { isDeepStrictEqual } = require("../../util/Utils");
class SharedUser extends Base {
    constructor(client, data) {
        super(client);
        this.userId = String(data.user_id);
        this._patch(data);
    }
    _patch(data) {
        if ("first_name" in data) {
            this.firstName = data.first_name;
        }
        if ("last_name" in data) {
            this.lastName = data.last_name;
        }
        if ("username" in data) {
            this.username = data.username;
        }
        if ("photo" in data) {
            this.photo = data.photo.map((photo) => new Photo(this.client, photo));
        }
        return data;
    }
    fetch({ force = true, fullInfo = false } = {}) {
        return this.client.users.fetch(this.userId, { force, fullInfo });
    }
    send(text, options = {}) {
        if (typeof text === "object") {
            return this.client.sendMediaGroup({
                chatId: this.userId,
                ...text,
            });
        }
        return this.client.sendMessage({
            text,
            chatId: this.userId,
            ...options,
        });
    }
    sendGift(giftId, options = {}) {
        return this.client.sendGift({
            giftId,
            userId: this.userId,
            ...options,
        });
    }
    giftPremiumSubscription(monthCount, starCount, options = {}) {
        return this.client.giftPremiumSubscription({
            ...options,
            monthCount,
            starCount,
            userId: this.userId,
        });
    }
    saveInlineMessage(result, options = {}) {
        return this.client.savePreparedInlineMessage({
            result,
            userId: this.userId,
            ...options,
        });
    }
    setStarSubscription(telegramPaymentChargeId, isCanceled) {
        return this.client.editUserStarSubscription({
            userId: this.userId,
            telegramPaymentChargeId,
            isCanceled,
        });
    }
    refundStarPayment(telegramPaymentId) {
        return this.client.refundStarPayment(this.userId, telegramPaymentId);
    }
    setPassportErrors(errors) {
        return this.client.setPassportDataErrors(this.userId, errors);
    }
    fetchProfilePhotos(offset = 0, limit = 100) {
        return this.client.getUserProfilePhotos({
            userId: this.userId,
            limit,
            offset,
        });
    }
    fetchChatBoosts(chatId) {
        return this.client.getUserChatBoosts(chatId, this.userId);
    }
    setEmojiStatus({ emojiStatusCustomEmojiId, emojiStatusExpirationDate } = {}) {
        return this.client.setUserEmojiStatus({
            userId: this.userId,
            ...(emojiStatusCustomEmojiId && { emojiStatusCustomEmojiId }),
            ...(emojiStatusExpirationDate && { emojiStatusExpirationDate }),
        });
    }
    verify(description) {
        return this.client.verifyUser(this.userId, description);
    }
    removeVerification() {
        return this.client.removeUserVerification(this.userId);
    }
    equals(other) {
        if (!other || !(other instanceof SharedUser))
            return false;
        return (this.userId === other.userId &&
            this.firstName === other.firstName &&
            this.lastName === other.lastName &&
            this.username === other.username &&
            isDeepStrictEqual(this.photo, other.photo));
    }
    toString() {
        var _a;
        return (_a = this.username) !== null && _a !== void 0 ? _a : "";
    }
}
module.exports = { SharedUser };
