// telegramsjs/dist/src/structures/CallbackQuery.js
"use strict";
const { Base } = require("./Base");
const { Message } = require("./message/Message");
class CallbackQuery extends Base {
    constructor(client, data) {
        super(client);
        this.id = data.id;
        this.author = this.client.users._add(data.from);
        if ("message" in data) {
            this.message = new Message(client, data.message);
        }
        if ("inline_message_id" in data) {
            this.inlineMessageId = data.inline_message_id;
        }
        this.chatInstance = data.chat_instance;
        if ("data" in data) {
            this.data = data.data;
        }
        if ("game_short_name" in data) {
            this.gameShortName = data.game_short_name;
        }
    }
    send(text, options = {}) {
        return this.client.answerCallbackQuery({
            callbackQueryId: this.id,
            text,
            ...options,
        });
    }
    showAlert(text, url, cacheTime) {
        return this.client.answerCallbackQuery({
            callbackQueryId: this.id,
            showAlert: true,
            text,
            ...(url && { url }),
            ...(cacheTime && { cacheTime }),
        });
    }
}
module.exports = { CallbackQuery };
