// telegramsjs/dist/src/structures/ChatJoinRequest.js
"use strict";
const { Base } = require("./Base");
const { ChatInviteLink } = require("./chat/ChatInviteLink");
class ChatJoinRequest extends Base {
    constructor(client, data) {
        super(client);
        this.userChatId = String(data.user_chat_id);
        this.chat = this.client.chats._add(data.chat);
        this.author = this.client.users._add(data.from);
        if ("bio" in data) {
            this.bio = data.bio;
        }
        if ("invite_link" in data) {
            this.inviteLink = new ChatInviteLink(client, data.invite_link);
        }
        this.createdUnixTime = data.date;
    }
    get createdTimestamp() {
        return this.createdUnixTime * 1000;
    }
    get createdAt() {
        return new Date(this.createdTimestamp);
    }
    approveJoinRequest() {
        return this.client.approveChatJoinRequest(this.chat.id, this.author.id);
    }
    declineJoinRequest() {
        return this.client.declineChatJoinRequest(this.chat.id, this.author.id);
    }
}
module.exports = { ChatJoinRequest };
