// telegramsjs/dist/src/structures/ChatMemberUpdated.js
"use strict";
const { Base } = require("./Base");
const { ChatMember } = require("./chat/ChatMember");
const { ChatInviteLink } = require("./chat/ChatInviteLink");
class ChatMemberUpdated extends Base {
    constructor(client, data) {
        super(client);
        this.chat = this.client.chats._add(data.chat);
        this.author = this.client.users._add(data.from);
        this.createdUnixTime = data.date;
        this.oldMember = new ChatMember(client, this.chat.id, data.old_chat_member);
        this.newMember = new ChatMember(client, this.chat.id, data.new_chat_member);
        if ("invite_link" in data) {
            this.inviteLink = new ChatInviteLink(client, data.invite_link);
        }
        if ("via_join_request" in data) {
            this.viaJoinRequest = data.via_join_request;
        }
        if ("via_chat_folder_invite_link" in data) {
            this.viaInviteLink = data.via_chat_folder_invite_link;
        }
    }
    get createdTimestamp() {
        return this.createdUnixTime * 1000;
    }
    get createdAt() {
        return new Date(this.createdTimestamp);
    }
}
module.exports = { ChatMemberUpdated };
