// telegramsjs/dist/src/structures/ShippingQuery.js
"use strict";
const { Base } = require("./Base");
class ShippingQuery extends Base {
    constructor(client, data) {
        var _a, _b, _c, _d, _e, _f;
        super(client);
        this.id = data.id;
        this.author = this.client.users._add(data.from);
        this.invoicePayload = data.invoice_payload;
        const shippingAddress = {};
        shippingAddress.countryCode = (_a = data.shipping_address) === null || _a === void 0 ? void 0 : _a.country_code;
        shippingAddress.state = (_b = data.shipping_address) === null || _b === void 0 ? void 0 : _b.state;
        shippingAddress.city = (_c = data.shipping_address) === null || _c === void 0 ? void 0 : _c.city;
        shippingAddress.streetLine1 = (_d = data.shipping_address) === null || _d === void 0 ? void 0 : _d.street_line1;
        shippingAddress.streetLine2 = (_e = data.shipping_address) === null || _e === void 0 ? void 0 : _e.street_line2;
        shippingAddress.postCode = (_f = data.shipping_address) === null || _f === void 0 ? void 0 : _f.post_code;
        this.shippingAddress = shippingAddress;
    }
    answerQuery(ok, options = {}) {
        return this.client.answerShippingQuery({
            shippingQueryId: this.id,
            ok,
            ...options,
        });
    }
}
module.exports = { ShippingQuery };
