// telegramsjs/dist/src/structures/gift/Gift.js
"use strict";
const { Base } = require("../Base");
const { Sticker } = require("../media/Sticker");
const { isDeepStrictEqual } = require("../../util/Utils");
class Gift extends Base {
    constructor(client, data) {
        super(client);
        this.id = data.id;
        this.sticker = new Sticker(client, data.sticker);
        this.startCount = data.star_count;
        if ("upgrade_star_count" in data) {
            this.upgradeStarCount = data.upgrade_star_count;
        }
        if ("total_count" in data) {
            this.totalCount = data.total_count;
        }
        if ("remaining_count" in data) {
            this.remainingCount = data.remaining_count;
        }
    }
    sendGift(userId, options = {}) {
        return this.client.sendGift({
            userId,
            giftId: this.id,
            ...options,
        });
    }
    equals(other) {
        if (!other || !(other instanceof Gift))
            return false;
        return (this.id === other.id &&
            isDeepStrictEqual(this.sticker, other.sticker) &&
            this.startCount === other.startCount &&
            this.totalCount === other.totalCount &&
            this.upgradeStarCount === other.upgradeStarCount &&
            this.remainingCount === other.remainingCount);
    }
}
module.exports = { Gift };
