// telegramsjs/dist/src/structures/gift/GiftInfo.js
"use strict";
const { Base } = require("../Base");
const { Gift } = require("./Gift");
const { MessageEntities } = require("../message/MessageEntities");
class GiftInfo extends Base {
    constructor(client, data) {
        super(client);
        this.gift = new Gift(client, data.gift);
        if ("owned_gift_id" in data) {
            this.ownedGiftId = data.owned_gift_id;
        }
        if ("text" in data) {
            this.content = data.text;
            if ("entities" in data) {
                this.entities = new MessageEntities(client, data.text, data.entities);
            }
        }
        if ("is_private" in data) {
            this.isPrivate = data.is_private;
        }
        if ("can_be_upgraded" in data) {
            this.beUpgraded = data.can_be_upgraded;
        }
        if ("convert_star_count" in data) {
            this.convertStarCount = data.convert_star_count;
        }
        if ("prepaid_upgrade_star_count" in data) {
            this.prepaidUpgradeStarCount = data.prepaid_upgrade_star_count;
        }
    }
}
module.exports = { GiftInfo };
