// telegramsjs/dist/src/structures/gift/OwnedGiftUnique.js
"use strict";
const { Base } = require("../Base");
const { UniqueGift } = require("./UniqueGift");
class OwnedGiftUnique extends Base {
    constructor(client, data) {
        super(client);
        this.type = data.type;
        this.gift = new UniqueGift(client, data.gift);
        if ("owned_gift_id" in data) {
            this.ownedGiftId = data.owned_gift_id;
        }
        if ("sender_user" in data) {
            this.senderUser = client.users._add(data.sender_user);
        }
        this.senderUnixTime = data.send_date;
        if ("is_saved" in data) {
            this.isSaved = data.is_saved;
        }
        if ("can_be_transferred" in data) {
            this.beTransferred = data.can_be_transferred;
        }
        if ("transfer_star_count" in data) {
            this.transferStarCount = data.transfer_star_count;
        }
    }
    get senderTimestamp() {
        return this.senderUnixTime * 1000;
    }
    get senderAt() {
        return new Date(this.senderUnixTime);
    }
}
module.exports = { OwnedGiftUnique };
