// telegramsjs/dist/src/structures/gift/UniqueGift.js
"use strict";
const { Base } = require("../Base");
const { Sticker } = require("../media/Sticker");
class UniqueGift extends Base {
    constructor(client, data) {
        super(client);
        this.baseName = data.base_name;
        this.name = data.name;
        this.number = data.number;
        this.model = {
            name: data.model.name,
            sticker: new Sticker(client, data.model.sticker),
            rarityPerMille: data.model.rarity_per_mille,
        };
        this.symbol = {
            name: data.symbol.name,
            sticker: new Sticker(client, data.symbol.sticker),
            rarityPerMille: data.symbol.rarity_per_mille,
        };
        this.backdrop = {
            name: data.backdrop.name,
            colors: {
                center: data.backdrop.colors.center_color,
                edge: data.backdrop.colors.edge_color,
                symbol: data.backdrop.colors.symbol_color,
                text: data.backdrop.colors.text_color,
            },
            rarityPerMille: data.backdrop.rarity_per_mille,
        };
    }
}
module.exports = { UniqueGift };
