// telegramsjs/dist/src/rest/MultipartStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipartStream = void 0;
const sandwich_stream_1 = require("sandwich-stream");
const node_stream_1 = require("node:stream");
class MultipartStream extends sandwich_stream_1.SandwichStream {
    constructor(boundary) {
        super({
            head: `--${boundary}\r\n`,
            tail: `\r\n--${boundary}--`,
            separator: `\r\n--${boundary}\r\n`,
        });
    }
    addPart(part = {}) {
        const partStream = new node_stream_1.PassThrough();
        if (part.headers) {
            for (const key in part.headers) {
                const header = part.headers[key];
                partStream.write(`${key}:${header}\r\n`);
            }
        }
        partStream.write("\r\n");
        if (MultipartStream.isStream(part.body)) {
            part.body.pipe(partStream);
        }
        else {
            partStream.end(part.body);
        }
        this.add(partStream);
    }
    static isStream(stream) {
        return (stream && typeof stream === "object" && typeof (stream === null || stream === void 0 ? void 0 : stream.pipe) === "function");
    }
}
exports.MultipartStream = MultipartStream;
