// telegramsjs/dist/src/util/inline/InlineQueryResultCachedBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineQueryResultCachedBuilder = void 0;
class InlineQueryResultCachedBuilder {
    static audio(id, fileId, options) {
        return { type: "audio", id, audio_file_id: fileId, ...options };
    }
    static document(id, title, fileId, options) {
        return {
            type: "document",
            id,
            title,
            document_file_id: fileId,
            ...options,
        };
    }
    static gif(id, fileId, options) {
        return { type: "gif", id, gif_file_id: fileId, ...options };
    }
    static mpeg4Gif(id, fileId, options) {
        return { type: "mpeg4_gif", id, mpeg4_file_id: fileId, ...options };
    }
    static photo(id, fileId, options) {
        return { type: "photo", id, photo_file_id: fileId, ...options };
    }
    static sticker(id, fileId, options) {
        return { type: "sticker", id, sticker_file_id: fileId, ...options };
    }
    static video(id, title, fileId, options) {
        return {
            type: "video",
            id,
            title,
            video_file_id: fileId,
            ...options,
        };
    }
    static voice(id, title, fileId, options) {
        return {
            type: "voice",
            id,
            title,
            voice_file_id: fileId,
            ...options,
        };
    }
}
exports.InlineQueryResultCachedBuilder = InlineQueryResultCachedBuilder;
