// telegramsjs/dist/src/errors/ErrorCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorMessages = exports.ErrorCodes = void 0;
var ErrorCodes;
(function (ErrorCodes) {
    ErrorCodes["InvalidOptions"] = "INVALID_OPTIONS";
    ErrorCodes["MissingUrlParameter"] = "MISSING_URL_PARAMETER";
    ErrorCodes["MissingToken"] = "MISSING_TOKEN";
    ErrorCodes["WebhookServerCreationFailed"] = "WEBHOOK_SERVER_CREATION_FAILED";
    ErrorCodes["InvalidFilterFunction"] = "INVALID_FILTER_FUNCTION";
    ErrorCodes["UserIdNotAvailable"] = "USER_ID_NOT_AVAILABLE";
    ErrorCodes["MessageIdNotAvailable"] = "MESSAGE_ID_NOT_AVAILABLE";
    ErrorCodes["ChatIdNotAvailable"] = "CHAT_ID_NOT_AVAILABLE";
    ErrorCodes["FileRetrievalFailed"] = "FILE_RETRIEVAL_FAILED";
    ErrorCodes["FileDownloadFailed"] = "FILE_DOWNLOAD_FAILED";
    ErrorCodes["FileWriteInvalidType"] = "FILE_WRITE_INVALID_TYPE";
    ErrorCodes["InvalidUserId"] = "INVALID_USER_ID";
    ErrorCodes["InvalidChatId"] = "INVALID_CHAT_ID";
    ErrorCodes["InvalidClientId"] = "INVALID_CLIENT_ID";
    ErrorCodes["InvalidFileName"] = "INVALID_FILE_NAME";
})(ErrorCodes || (exports.ErrorCodes = ErrorCodes = {}));
const ErrorMessages = {
    [ErrorCodes.InvalidOptions]: "The provided options are invalid.",
    [ErrorCodes.MissingUrlParameter]: "The 'url' parameter is required but was not provided.",
    [ErrorCodes.MissingToken]: "A token must be specified to receive updates from Telegram.",
    [ErrorCodes.WebhookServerCreationFailed]: "The webhook server could not be created.",
    [ErrorCodes.InvalidFilterFunction]: "The provided 'options.filter' is not a function.",
    [ErrorCodes.UserIdNotAvailable]: "The user ID related to this message is not available.",
    [ErrorCodes.MessageIdNotAvailable]: "The message ID related to this message is not available.",
    [ErrorCodes.ChatIdNotAvailable]: "The chat ID related to this message is not available.",
    [ErrorCodes.FileRetrievalFailed]: "Failed to retrieve the file from the path: <file_path>.",
    [ErrorCodes.FileDownloadFailed]: "Failed to download the file. Error: ${err}.",
    [ErrorCodes.FileWriteInvalidType]: "Invalid file write type specified. Available types: 'stream' or 'promise'.",
    [ErrorCodes.InvalidUserId]: "The provided ID is invalid for retrieving user information; it does not correspond to a valid user ID.",
    [ErrorCodes.InvalidChatId]: "The provided ID is invalid for retrieving chat information; it does not correspond to a valid chat ID.",
    [ErrorCodes.InvalidClientId]: "The bot ID is not available. Please check if the bot has been initialized",
    [ErrorCodes.InvalidFileName]: "The name file is not valid. Please open issue https://github.com/telegramsjs/Telegramsjs/issues",
};
exports.ErrorMessages = ErrorMessages;
