// @telegram.ts/formatters/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedLanguages = exports.version = exports.markdownV2Patterns = exports.markdownPatterns = exports.htmlEntityPatterns = exports.escapeMarkdown = exports.escapeHTML = exports.parseEntities = exports.markdownv = exports.html = void 0;
const html_1 = require("./html");
exports.html = html_1.default;
const markdownv_1 = require("./markdownv");
exports.markdownv = markdownv_1.default;
const utils_1 = require("./utils");
Object.defineProperty(exports, "parseEntities", { enumerable: true, get: function () { return utils_1.parseEntities; } });
Object.defineProperty(exports, "escapeHTML", { enumerable: true, get: function () { return utils_1.escapeHTML; } });
Object.defineProperty(exports, "escapeMarkdown", { enumerable: true, get: function () { return utils_1.escapeMarkdown; } });
Object.defineProperty(exports, "htmlEntityPatterns", { enumerable: true, get: function () { return utils_1.htmlEntityPatterns; } });
Object.defineProperty(exports, "markdownPatterns", { enumerable: true, get: function () { return utils_1.markdownPatterns; } });
Object.defineProperty(exports, "markdownV2Patterns", { enumerable: true, get: function () { return utils_1.markdownV2Patterns; } });
const package_json_1 = require("../package.json");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return package_json_1.version; } });
const constants_1 = require("./constants");
Object.defineProperty(exports, "SupportedLanguages", { enumerable: true, get: function () { return constants_1.SupportedLanguages; } });
exports.default = {
    html: html_1.default,
    markdownv: markdownv_1.default,
    parseEntities: utils_1.parseEntities,
    escapeHTML: utils_1.escapeHTML,
    escapeMarkdown: utils_1.escapeMarkdown,
    htmlEntityPatterns: utils_1.htmlEntityPatterns,
    markdownPatterns: utils_1.markdownPatterns,
    markdownV2Patterns: utils_1.markdownV2Patterns,
    version: package_json_1.version,
    SupportedLanguages: constants_1.SupportedLanguages,
};
