// telegramsjs/dist/src/util/permission/BusinessPermissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BusinessPermissions = void 0;
class BusinessPermissions {
    constructor(data = {}) {
        Object.defineProperty(this, "allowed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "denied", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.allowed = new Set();
        this.denied = new Set();
        this._patch(data);
    }
    allow(permissions) {
        if (!permissions)
            return this;
        if (permissions instanceof BusinessPermissions) {
            permissions = permissions.toObject();
        }
        else if (typeof permissions === "string") {
            permissions = { [permissions]: true };
        }
        for (const [key, value] of Object.entries(permissions)) {
            const perm = key;
            if (value) {
                this.allowed.add(perm);
                this.denied.delete(perm);
            }
        }
        return this;
    }
    deny(permissions) {
        if (!permissions)
            return this;
        if (permissions instanceof BusinessPermissions) {
            permissions = permissions.toObject();
        }
        else if (typeof permissions === "string") {
            permissions = { [permissions]: true };
        }
        for (const [key, value] of Object.entries(permissions)) {
            const perm = key;
            if (value) {
                this.denied.add(perm);
                this.allowed.delete(perm);
            }
        }
        return this;
    }
    has(permission) {
        return this.allowed.has(permission);
    }
    toObject() {
        const flags = {};
        for (const perm of this.allowed) {
            flags[perm] = true;
        }
        for (const perm of this.denied) {
            if (!flags[perm]) {
                flags[perm] = false;
            }
        }
        return flags;
    }
    equals(other) {
        if (!other || !(other instanceof BusinessPermissions))
            return false;
        const thisAllowed = Array.from(this.allowed).sort();
        const otherAllowed = Array.from(other.allowed).sort();
        const thisDenied = Array.from(this.denied).sort();
        const otherDenied = Array.from(other.denied).sort();
        return (thisAllowed.length === otherAllowed.length &&
            thisDenied.length === otherDenied.length &&
            thisAllowed.every((perm, index) => perm === otherAllowed[index]) &&
            thisDenied.every((perm, index) => perm === otherDenied[index]));
    }
    _patch(data) {
        for (const [key, value] of Object.entries(data)) {
            if (value) {
                this.allowed.add(key);
            }
            else {
                this.denied.add(key);
            }
        }
    }
    static isValid(permission) {
        return Object.keys(BusinessPermissions.Flags).includes(permission);
    }
}
exports.BusinessPermissions = BusinessPermissions;
Object.defineProperty(BusinessPermissions, "Flags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        canReply: 1,
        readMessages: 2,
        deleteOutgoingMessages: 3,
        deleteAllMessages: 4,
        editName: 5,
        editBio: 6,
        editProfilePhoto: 7,
        editUsername: 8,
        editStories: 9,
        changeGiftSettings: 10,
        viewGiftsAndStars: 11,
        postMessages: 12,
        convertGiftsToStars: 13,
        transferAndUpgradeGifts: 14,
        transferStars: 15,
        manageStories: 16,
    }
});
